app/test/unit/git/squash-test.ts

```diff
@@ -34,18 +45,269 @@ describe('git/cherry-pick', () => {
+    const secondCommit = await makeSquashCommit(repository, 'second')
+    const thirdCommit = await makeSquashCommit(repository, 'third')
+    const fourthCommit = await makeSquashCommit(repository, 'fourth')
+
+    const result = await squash(
+      repository,
+      [secondCommit, thirdCommit, fourthCommit],
+      firstCommit,
+      initialCommit.sha,
+      'Test Summary\n\nTest Body'
+    )
+
+    expect(result).toBe(RebaseResult.CompletedWithoutError)
+
+    const log = await getCommits(repository, 'HEAD', 5)
+    const squashed = log[0]
+    expect(squashed.summary).toBe('Test Summary')
+    expect(squashed.body).toBe('Test Body\n')
+    expect(log.length).toBe(2)
+
+    // verify squashed commit contains changes from squashed commits
+    const squashedFiles = await getChangedFiles(repository, squashed.sha)
+    const squashedFilePaths = squashedFiles.map(f => f.path).join(' ')
+    expect(squashedFilePaths.includes('first.md')).toBeTrue()
+    expect(squashedFilePaths.includes('second.md')).toBeTrue()
+    expect(squashedFilePaths.includes('third.md')).toBeTrue()
+    expect(squashedFilePaths.includes('fourth.md')).toBeTrue()
+  })
+
+  it('squashes multiple commit non-sequential commits (reorders, non-conflicting)', async () => {
+    const firstCommit = await makeSquashCommit(repository, 'first')
+    const secondCommit = await makeSquashCommit(repository, 'second')
+    await makeSquashCommit(repository, 'third')
+    const fourthCommit = await makeSquashCommit(repository, 'fourth')
+
+    // From oldest to newest, log looks like:
+    // - initial commit
+    // - 'first' commit
+    // - 'second' commit
+    // - 'third' commit
+    // - 'fourth' commit
+
+    // Squashing 'second' and 'fourth' onto 'first'
+    // Thus, reordering 'fourth' to be before 'third'
+    const result = await squash(
+      repository,
+      [secondCommit, fourthCommit],
+      firstCommit,
+      initialCommit.sha,
+      'Test Summary\n\nTest Body'
+    )
+
+    expect(result).toBe(RebaseResult.CompletedWithoutError)
+
+    // From oldest to newest, log should look like:
+    // - initial commit - log[2]
+    // - the squashed commit 'Test Summary' - log[1]
+    // - 'third' commit - log[0]
+    const log = await getCommits(repository, 'HEAD', 5)
+    const squashed = log[1]
+    expect(squashed.summary).toBe('Test Summary')
+    expect(squashed.body).toBe('Test Body\n')
+    expect(log[0].summary).toBe('third')
+    expect(log.length).toBe(3)
+
+    // verify squashed commit contains changes from squashed commits
+    const squashedFiles = await getChangedFiles(repository, squashed.sha)
+    const squashedFilePaths = squashedFiles.map(f => f.path).join(' ')
+    expect(squashedFilePaths.includes('first.md')).toBeTrue()
+    expect(squashedFilePaths.includes('second.md')).toBeTrue()
+    expect(squashedFilePaths.includes('fourth.md')).toBeTrue()
+    expect(squashedFilePaths.includes('third.md')).toBeFalse()
```

Comment:
Same (but `third.md` has a `not`):
```suggestion
    expect(squashedFilePaths).toContain('first.md')
    expect(squashedFilePaths).toContain('second.md')
    expect(squashedFilePaths).toContain('fourth.md')
    expect(squashedFilePaths).not.toContain('third.md')
```